### ISSUE TITLE
Wifi scan results in incorrect/garbage network name, makes wifi unusable

### ISSUE THREAD

**Issue by joshuagiles** [libretro/RetroArch#7227]
## tl;dr

Undefined behavior in connmanctl wifi "driver" results in garbage wifi SSID names, relevant code linked in the long text

## The Bug

I discovered this bug when I was trying to install Lakka on a laptop (amd64). The first time I booted it up I went to the wifi option and it scanned for wifi networks and found several, as expected. However I booted it up a second time (and several more times) and each time the wifi scan populated the list of wifi networks with just one entry, and that one entry had a name consisting of seemingly random unicode characters (different every time). Boxes, symbols, just garbage. Trying to connect to this network results in an error message `Error "connect" invalid argument` (I apologize, I'm paraphrasing as this message pops up on-screen and can't be copy-pasted, and it's now been several minutes since I last saw it).

I looked around online and found this issue has definitely plagued several other people
* this Lakka issue, closed with no action https://github.com/libretro/Lakka-LibreELEC/issues/325
* this forum post https://forums.libretro.com/t/wifi-orange-pi-lite-not-connect/15638
* this forum reply also mentions the issue https://forums.libretro.com/t/error-enabling-wifi-on-pc/4786/12

I decided to try a nightly build from nightly.builds.lakka.tv, and found it has the same issue (I used the build from Sep 14, I don't know which particular version of retroarch it uses, but it is not relevant as the offending code is still in the retroarch git repo, untouched for 2 years).

At this point, not ready to be defeated yet, I dove into the retroarch source code here on github, to find the part of the source code which was being executed to produce that garbage SSID, so I could see the exact connmanctl commands it was invoking, so I could do them myself in a terminal and see what the problem was.

And there we are: https://github.com/libretro/RetroArch/blob/master/wifi/drivers/connmanctl.c#L61

I found that when I booted the second time into a terminal, and used those commands myself, I got the following results:

`connmanctl enable wifi`: Error wifi: Already enabled
`connmanctl scan wifi`: Wifi scan complete.
`connmanctl services`: *nothing but a single newline character*

There it is. The connmanctl functions in retroarch do not consider the possibility that that command might return anything other than a full list of scanned networks, in the expected format. That string "\n" gets turned into "\0" in the following loop, and appended to the list of SSIDs. Later when retroarch calls to `connmanctl_get_ssids`, this line (https://github.com/libretro/RetroArch/blob/master/wifi/drivers/connmanctl.c#L93) takes the pointer to that string "\0", with length 1, and adds 4 to it. Based on seeing the *expected* output from `connmanctl services`, this is to skip past the four-character preamble on each line, before the SSID is listed. Because the possibility that the output ends before even that isn't handled, this results in the SSID name being read from memory which is three bytes *past the end* of the string allocation, resulting in those garbage SSID names in the GUI menu.

## The Fix?

In my own very short experiments, I found that I could solve the issue when running the commands myself by simply disabling and then re-enabling wifi, and then re-scanning. (`connmanctl disable wifi; connmanctl enable wifi; connmanctl scan wifi; connmanctl services`)

I believe that, besides making the scan function account for the possibility of no scan results (maybe even automatically disable/re-enable to re-scan once, just in case?), this problem could also be made much easier for end-users to handle themselves, by adding GUI elements allowing the user to enable, disable, and re-scan for wifi networks if they choose to. As it is, the GUI allows the user to scan once, when they select the wifi option, and then uses that list of networks forever afterward. Quite awkward if you have to do any sort of wifi troubleshooting.

Hope this helps you folks fix this bug in time for the next release, it really is a game-breaker for some potential retroarch (and Lakka) users like myself who cannot rely on having an ethernet cable plugged into the device running the software.

**Comment by bparker06** [comment]
You might want to create an issue on the lakka repo for this, since the wifi support is lakka-only and would probably have more visibility over there with people that would want to fix it.

**Comment by metaldudu** [comment]
after install lakka 2.2 in my old netbook eeepc 1000H, i found same error. In the wifi menu, only show one random character, and can not connect it.
In command line, i also could scan wifi. maybe next version could fix this bug.

**Comment by papou84** [comment]
Same issue after trying to set the correct time zone to Europe/Paris.Obviously it is impossible to connect with ssh.

**Comment by matcarfer** [comment]
Tested latests nightlies for last few months and can confirm this issue doesnt happen to me, what ir does happen (and have reported on discord many times) is that when tryong to connect, despite making the connection correct (sometimes yo have to try more than one time), it does give some errors. It still works, but some errors show up.
